README.md
# Blist theme for Hugo

![Latest Release](https://img.shields.io/github/tag/apvarun/blist-hugo-theme.svg)
![Blist hugo theme](https://img.shields.io/github/license/apvarun/blist-hugo-theme)
![Hugo generator](https://img.shields.io/badge/generator-hugo-brightgreen)

Blist is a clean and fast blog theme for your Hugo site.

![Blist Icon](https://github.com/apvarun/blist-hugo-theme/raw/main/images/blist-logo.png)

**Features:**

- Responsive content / Mobile-optimized
- Blog pagination
- Customizable Text Search
- Table of contents
- Social links
- Code highlighting
- Color customization
- Dark mode
- Fast performance
- SEO optimized
- i18n support
- Support comments systems
  - Disqus (hugo native)
  - giscus
  - utterances

## Preview

![Preview](https://github.com/apvarun/blist-hugo-theme/raw/main/images/screenshot.png)

➡️ [DEMO](https://blist.vercel.app/)

## Get the theme

Minimum Hugo Version: **0.82.1**

Run from the root of your Hugo site:

```sh
git clone https://github.com/apvarun/blist-hugo-theme.git themes/blist
```

Alternatively, you can include this repository as a [git submodule](https://git-scm.com/docs/gitsubmodules). This makes it easier to update this theme if you have your Hugo site in git as well:

```sh
git submodule add https://github.com/apvarun/blist-hugo-theme.git themes/blist
```

## Preview the theme with example content

Blist theme ships with an fully configured example site. For a quick preview:

Copy the `package.json` file from `themes/blist` folder to your hugo website root folder, and run `npm install`.

```sh
cd themes/blist/exampleSite/
hugo serve --themesDir ../..
```

Then visit `http://localhost:1313/` in your browser to view the example site.

## Configuring theme to a hugo website

1. Copy `package.json` and `package-lock.json` to the root folder of your the website
2. Run `npm install` to install required packages for theme
3. Run `npm i -g postcss-cli` to use PostCSS with Hugo build
4. Set `theme = 'blist'` in config.toml
5. Run `npm start` to start your local server

Make sure to commit the above changes to your repository.

## Publish your website

When deploying to services like Netlify or Vercel, use the following command for building your site:

```sh
npm i && HUGO_ENVIRONMENT=production hugo --gc
```
The parameter `HUGO_ENVIRONMENT=production` enables the execution of css purging.

## Add content

The following explains how to add content to your Hugo site. You can find sample content in the `exampleSite/` folder.

### Structure:

    .
    ├── ...
    ├── blog       # Blog Section
    │   ├── post1   # Post 1
    │   ├── post2   # Post 2
    │   └── _index
    └── ...

## Configure your site

From `exampleSite/`, copy `config.toml` to the root folder of your Hugo site and change the fields as you like. Helpful comments are provided.

### Menu

Menu in Blist theme is pre-set to have all section names. You can include custom links in header using the `menu.main` option config.toml.

### Logo

`logo` param in the site config will allow to use an image as the logo instead of the website name. It is localizable and so can have different logo for different languages

### Darkmode

`[params.darkModeToggle]` enables the dark mode toggle in header. The preference is then saved so that the mode is automatically chosen for return visits.

### Customize Ascent Color

Use `[params.ascentColor]` to change the default `pink` color to any supported color from the [list of default colors](https://tailwindcss.com/docs/customizing-colors) from Tailwind CSS.

Some example values: bg-blue-200, bg-yellow-300

### Search

`[params.enableSearch]` option is used to enable search option in the theme.

- Adds the search icon in header
- Generates the search index
- Uses fuse.js to enable searching through content

In order to search, you can either click on the search icon from header or press `Ctrl/Cmd + /` key combination.

**Note:**

Make sure to enable JSON in outputs array.

```
[outputs]
  home = ["HTML", "RSS", "JSON"]
```

### Latex

Enable Mathematical options: set `math: true` in your markdown frontmatter

### Google Analytics

Set `googleAnalytics` in `config.toml` to activate Hugo's [internal Google Analytics template](https://gohugo.io/templates/internal/#google-analytics).

## Performance

[![Pagespeed Insights Performance](https://github.com/apvarun/blist-hugo-theme/raw/main/images/pagespeed-performance.png)](https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fblist.vercel.app&tab=mobile)

## Issues

If you have a question, please [open an issue](https://github.com/apvarun/blist-hugo-theme/issues) for help and to help those who come after you. The more information you can provide, the better!

## Contributing

Contributions, issues, and feature requests are welcome! For major changes, please open an issue first to discuss what you would like to change.

<table>
  <tr>
    <td align="center"><a href="https://github.com/apvarun"><img alt="apvarun"
          src="https://avatars.githubusercontent.com/u/8411309?v=4" width="117" /><br />apvarun</a></td>
    <td align="center"><a href="https://github.com/Ravinou"><img alt="Ravinou"
          src="https://avatars.githubusercontent.com/u/39600829?v=4" width="117" /><br />Ravinou</a></td>
    <td align="center"><a href="https://github.com/engel-b"><img alt="engel-b"
          src="https://avatars.githubusercontent.com/u/5812810?v=4" width="117" /><br />engel-b</a></td>
    <td align="center"><a href="https://github.com/chris-gillatt"><img alt="chris-gillatt"
          src="https://avatars.githubusercontent.com/u/9963006?v=4" width="117" /><br />chris-gillatt</a></td>
    <td align="center"><a href="https://github.com/apps/dependabot"><img alt="dependabot[bot]"
          src="https://avatars.githubusercontent.com/in/29110?v=4" width="117" /><br />dependabot[bot]</a></td>
    <td align="center"><a href="https://github.com/xLexip"><img alt="xLexip"
          src="https://avatars.githubusercontent.com/u/46321162?v=4" width="117" /><br />xLexip</a></td>
  </tr>
  <tr>
    <td align="center"><a href="https://github.com/AntoineSoetewey"><img alt="AntoineSoetewey"
          src="https://avatars.githubusercontent.com/u/17910063?v=4" width="117" /><br />AntoineSoetewey</a></td>
    <td align="center"><a href="https://github.com/chiragksharma"><img alt="chiragksharma"
          src="https://avatars.githubusercontent.com/u/74889769?v=4" width="117" /><br />chiragksharma</a></td>
    <td align="center"><a href="https://github.com/dlecan"><img alt="dlecan"
          src="https://avatars.githubusercontent.com/u/586631?v=4" width="117" /><br />dlecan</a></td>
    <td align="center"><a href="https://github.com/jpiaskowski"><img alt="jpiaskowski"
          src="https://avatars.githubusercontent.com/u/8420943?v=4" width="117" /><br />jpiaskowski</a></td>
    <td align="center"><a href="https://github.com/lucab85"><img alt="lucab85"
          src="https://avatars.githubusercontent.com/u/24267107?v=4" width="117" /><br />lucab85</a></td>
    <td align="center"><a href="https://github.com/Marty"><img alt="Marty"
          src="https://avatars.githubusercontent.com/u/100658?v=4" width="117" /><br />Marty</a></td>
  </tr>
  <tr>
    <td align="center"><a href="https://github.com/ernavdeepkr"><img alt="ernavdeepkr"
          src="https://avatars.githubusercontent.com/u/50796185?v=4" width="117" /><br />ernavdeepkr</a></td>
    <td align="center"><a href="https://github.com/roypeter"><img alt="roypeter"
          src="https://avatars.githubusercontent.com/u/16620459?v=4" width="117" /><br />roypeter</a></td>
  </tr>
</table>

## License

Licensed under [MIT](LICENSE)

## 🤝 Support

Give a ⭐️ if you like this project!

<a href="https://www.buymeacoffee.com/apvarun" target="_blank" rel="noopener"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" height="40" width="145" alt="Buy Me A Coffee"></a>


CHANGELOG.md
# Changelog

All the changes made to Blist theme for Hugo.

## v2.1.0 - 2023-01-09

### Added

- Support multiple comment systems #64

## v2.0.1 - 2022-12-27

### Fixed

- Page metadata layout in mobile #94

## v2.0.0 - 2022-12-25

### Added

- BREAKING: Ability to specify social icons order #91
- Table of contents support #84
- Search customization #79
- Option to skip showing thumbnail in post page #93
- Added more social icons
  - Mastodon #92
  - XDA #83
  - Telegram #83
  - Buy me a coffee #83
  - Google Scholar #85
  - Facebook #90
  - Behance #66
  - Tiktok #77
  - Stackoverflow #71

### Fixed

- Inline code block #78

## v1.5.0 - 2022-08-11

### Added

- Ability to hide metadata via param hideMeta #37
- Hide other languages option via param hideOtherLanguages
- Add French support #53 #54 #55

### Fixed

- Fix broken Multilingual search

## v1.4.0 - 2022-01-08

### Added

- Hamburger menu for mobile UI #23
- Image logo support #24

## v1.3.1 - 2021-11-28

### Added

- Adding xing support to social icons - #15
- Fix: Adding missing whitespace between label and language in single page - #16
- Fix: Readable Lastmod - #17

## v1.3.0 - 2021-10-27

### Added

- Add YouTube & Instagram social icons - #9
- Add i18n support - #11
- Add 404 page - #13

## v1.2.0 - 2021-07-18

### Added

- Add color customization for Intro and Social links blocks

## v1.1.0 - 2021-07-17

### Added

- Add support for text search

## v1.0.0 - 2021-07-16

- Initial Release

static/js/main.js
// Dean Attali / Beautiful Jekyll 2016

var main = {

  init : function() {
    // Shorten the navbar after scrolling a little bit down
    $(window).scroll(function() {
        if ($(".navbar").offset().top > 50) {
            $(".navbar").addClass("top-nav-short");
        } else {
            $(".navbar").removeClass("top-nav-short");
        }
    });

    // On mobile, hide the avatar when expanding the navbar menu
    $('#main-navbar').on('show.bs.collapse', function () {
      $(".navbar").addClass("top-nav-expanded");
    });
    $('#main-navbar').on('hidden.bs.collapse', function () {
      $(".navbar").removeClass("top-nav-expanded");
    });

    // On mobile, when clicking on a multi-level navbar menu, show the child links
    $('#main-navbar').on("click", ".navlinks-parent", function(e) {
      var target = e.target;
      $.each($(".navlinks-parent"), function(key, value) {
        if (value == target) {
          $(value).parent().toggleClass("show-children");
        } else {
          $(value).parent().removeClass("show-children");
        }
      });
    });

    // Ensure nested navbar menus are not longer than the menu header
    var menus = $(".navlinks-container");
    if (menus.length > 0) {
      var navbar = $("#main-navbar").find("ul");
      var fakeMenuHtml = "<li class='fake-menu' style='display:none;'><a></a></li>";
      navbar.append(fakeMenuHtml);
      var fakeMenu = $(".fake-menu");

      $.each(menus, function(i) {
        var parent = $(menus[i]).find(".navlinks-parent");
        var children = $(menus[i]).find(".navlinks-children a");
        var words = [];
        $.each(children, function(idx, el) { words = words.concat($(el).text().trim().split(/\s+/)); });
        var maxwidth = 0;
        $.each(words, function(id, word) {
          fakeMenu.html("<a>" + word + "</a>");
          var width =  fakeMenu.width();
          if (width > maxwidth) {
            maxwidth = width;
          }
        });
        $(menus[i]).css('min-width', maxwidth + 'px')
      });

      fakeMenu.remove();
    }
  },
};

// 2fc73a3a967e97599c9763d05e564189

document.addEventListener('DOMContentLoaded', main.init);

assets/css/tailwind.config.js
const themeDir = __dirname + "/../../";

module.exports = {
  purge: {
    enabled: process.env.HUGO_ENVIRONMENT === "production",
    content: [
      themeDir + "layouts/**/*.html",
      themeDir + "content/**/*.html",
      "layouts/**/*.html",
      "config.toml",
      "content/**/*.html",
      "assets/js/search.js",
      "exampleSite/layouts/**/*.html",
      "exampleSite/config.toml",
      "exampleSite/content/**/*.html",
    ],
  },
  darkMode: "class",
  theme: {
    extend: {
      typography: (theme) => ({
        DEFAULT: {
          css: [
            {
              'code::before': {
                content: '""',
              },
              'code::after': {
                content: '""',
              },
            }
          ]
        },
        dark: {
          css: [
            {
              color: theme("colors.gray.400"),
              '[class~="lead"]': {
                color: theme("colors.gray.300"),
              },
              a: {
                color: theme("colors.white"),
              },
              strong: {
                color: theme("colors.white"),
              },
              "ol > li::before": {
                color: theme("colors.gray.400"),
              },
              "ul > li::before": {
                backgroundColor: theme("colors.gray.600"),
              },
              hr: {
                borderColor: theme("colors.gray.200"),
              },
              blockquote: {
                color: theme("colors.gray.200"),
                borderLeftColor: theme("colors.gray.600"),
              },
              h1: {
                color: theme("colors.white"),
              },
              h2: {
                color: theme("colors.white"),
              },
              h3: {
                color: theme("colors.white"),
              },
              h4: {
                color: theme("colors.white"),
              },
              "figure figcaption": {
                color: theme("colors.gray.400"),
              },
              code: {
                color: theme("colors.white"),
              },
              "a code": {
                color: theme("colors.white"),
              },
              'code::before': {
                content: '""',
              },
              'code::after': {
                content: '""',
              },
              pre: {
                color: theme("colors.gray.200"),
                backgroundColor: theme("colors.gray.800"),
              },
              thead: {
                color: theme("colors.white"),
                borderBottomColor: theme("colors.gray.400"),
              },
              "tbody tr": {
                borderBottomColor: theme("colors.gray.600"),
              },
            },
          ],
        },
      }),
    },
  },
  variants: {
    extend: {
      typography: ["dark"],
    },
  },
  plugins: [require("@tailwindcss/typography")],
};


assets/js/search.js
// Credits to search implementation: https://gist.github.com/cmod/5410eae147e4318164258742dd053993

var fuse; // holds our search engine
var searchVisible = false;
var firstRun = true; // allow us to delay loading json data unless search activated
var list = document.querySelector('.search-list'); // targets the <ul>
var first = list.firstChild; // first child of search list
var last = list.lastChild; // last child of search list
var maininput = document.querySelector('.search-ui input'); // input box for search
var searchResultsHeading = document.querySelector('.search-results'); // input box for search
var noResults = document.querySelector('.no-results'); // input box for search
var resultsAvailable = false; // Did we get any search results?

// ==========================================
// The main keyboard event listener running the show
//
document.querySelector('.open-search').addEventListener('click', openSearch);
document.querySelector('.close-search').addEventListener('click', closeSearch);

function closeSearch() {
  document.querySelector('.search-ui').classList.add("hidden");
  document.activeElement.blur(); // remove focus from search box 
  searchVisible = false; // search not visible
  searchResultsHeading.classList.add('hidden');
}

function openSearch() {
  // Load json search index if first time invoking search
  // Means we don't load json unless searches are going to happen; keep user payload small unless needed
  if (firstRun) {
    loadSearch(); // loads our json data and builds fuse.js search index
    firstRun = false; // let's never do this again
  }

  //Close the mobile menu when search is click.
  mobileMenu.classList.toggle('hidden');

  // Toggle visibility of search box
  if (!searchVisible) {
    document.querySelector('.search-ui').classList.remove("hidden");
    document.querySelector('.search-ui input').focus(); // put focus in input box so you can just start typing
    searchVisible = true; // search visible
  }
}

document.addEventListener('keydown', function (event) {

  if (event.metaKey && event.which === 191) {
    openSearch()
  }

  // Allow ESC (27) to close search box
  if (event.keyCode == 27) {
    if (searchVisible) {
      document.querySelector('.search-ui').classList.add("hidden");
      document.activeElement.blur();
      searchVisible = false;
      searchResultsHeading.classList.add('hidden');
    }
  }

  // DOWN (40) arrow
  if (event.keyCode == 40) {
    if (searchVisible && resultsAvailable) {
      console.log("down");
      event.preventDefault(); // stop window from scrolling
      if (document.activeElement == maininput) { first.focus(); } // if the currently focused element is the main input --> focus the first <li>
      else if (document.activeElement == last) { last.focus(); } // if we're at the bottom, stay there
      else { document.activeElement.parentElement.nextSibling.firstElementChild.focus(); } // otherwise select the next search result
    }
  }

  // UP (38) arrow
  if (event.keyCode == 38) {
    if (searchVisible && resultsAvailable) {
      event.preventDefault(); // stop window from scrolling
      if (document.activeElement == maininput) { maininput.focus(); } // If we're in the input box, do nothing
      else if (document.activeElement == first) { maininput.focus(); } // If we're at the first item, go to input box
      else { document.activeElement.parentElement.previousSibling.firstElementChild.focus(); } // Otherwise, select the search result above the current active one
    }
  }
})


// ==========================================
// execute search as each character is typed
//
document.querySelector('.search-ui input').onkeyup = function (e) {
  executeSearch(this.value);
}


// ==========================================
// fetch some json without jquery
//
function fetchJSONFile(path, callback) {
  var httpRequest = new XMLHttpRequest();
  httpRequest.onreadystatechange = function () {
    if (httpRequest.readyState === 4) {
      if (httpRequest.status === 200) {
        var data = JSON.parse(httpRequest.responseText);
        if (callback) callback(data);
      }
    }
  };
  httpRequest.open('GET', path);
  httpRequest.send();
}


// ==========================================
// load our search index, only executed once
// on first call of search box (CMD-/)
//
function loadSearch() {
  const lang = document.querySelector('head > meta[name="lang"]')?.getAttribute?.('content')
  fetchJSONFile(`${lang ? "/" + lang : ""}/index.json`, function (data) {

    var options = { // fuse.js options; check fuse.js website for details
      shouldSort: true,
      location: 0,
      distance: 100,
      threshold: 0.4,
      minMatchCharLength: 2,
      keys: [
        'title',
        'permalink',
        'contents'
      ]
    };
    fuse = new Fuse(data, options); // build the index from the json file
  });
}


// ==========================================
// using the index we loaded on CMD-/, run 
// a search query (for "term") every time a letter is typed
// in the search box
//
function executeSearch(term) {
  let results = fuse.search(term); // the actual query being run using fuse.js
  let searchitems = ''; // our results bucket

  if (results.length === 0) { // no results based on what was typed into the input box
    resultsAvailable = false;
    searchitems = '';
    if (term !== "") {
      noResults.classList.remove('hidden')
    } else {
      noResults.classList.add('hidden')
    }
  } else { // build our html 
    noResults.classList.add('hidden')
    if (term !== "") {
      searchResultsHeading.classList.remove('hidden');
    }

    for (let item in results.slice(0, 5)) { // only show first 5 results
      const title = '<div class="text-2xl mb-2 font-bold">' + results[item].item.title + '</div>';
      const date = results[item].item.date ? '<div><em class="px-4">' + new Date(results[item].item.date).toUTCString().substring(0, 16) + '</em></div>' : '';
      const contents = '<div class="prose px-4">' + results[item].item.contents + '</div>';

      searchitems = searchitems + '<li><a class="block mb-2 px-4 py-2 rounded pb-2 border-b border-gray-200 dark:border-gray-600 focus:bg-gray-100 dark:focus:bg-gray-700 focus:outline-none" href="' + results[item].item.permalink + '" tabindex="0">' + title + '</a>' + date + contents + '</li>';
    }
    resultsAvailable = true;
  }

  list.innerHTML = searchitems;
  if (results.length > 0) {
    first = list.firstChild.firstElementChild; // first result container — used for checking against keyboard up/down location
    last = list.lastChild.firstElementChild; // last result container — used for checking against keyboard up/down location
  }
}
